src/utils.zig

```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
 	}
 };
 
+const ReadWriteTest = struct {
+	fn init() void {
+		main.initThreadLocals();
+	}
+	fn deinit() void {
+		main.deinitThreadLocals();
+	}
+	fn getWriter() BinaryWriter {
+		return .init(main.stackAllocator, .big);
+	}
+	fn getReader(data: []const u8) BinaryReader {
+		return .init(data, .big);
+	}
+	fn testInt(comptime intT: type, expected: intT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeInt(intT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readInt(intT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn testFloat(comptime floatT: type, expected: floatT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeFloat(floatT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readFloat(floatT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn testEnum(comptime enumT: type, expected: enumT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeEnum(enumT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readEnum(enumT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+	fn TestEnum(comptime intT: type) type {
+		return enum(intT) {
+			first = std.math.minInt(intT),
+			center = (std.math.maxInt(intT) + std.math.minInt(intT))/2,
+			last = std.math.maxInt(intT),
+		};
+	}
+	fn testVec(comptime vecT: type, expected: vecT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeVec(vecT, expected);
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readVec(vecT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+};
+
+test "read/write unsigned int" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	try ReadWriteTest.testInt(u1, 0);
+	try ReadWriteTest.testInt(u1, 1);
+
+	inline for([_]type{u2, u4, u5, u8, u16, u31, u32, u64, u128}) |intT| {
+		const max = std.math.maxInt(intT);
+		std.debug.assert(max != 0);
+
+		const mid = (std.math.maxInt(intT) + std.math.minInt(intT))/2;
+		std.debug.assert(mid != 0);
+
+		try ReadWriteTest.testInt(intT, std.math.minInt(intT));
+		try ReadWriteTest.testInt(intT, mid);
+		try ReadWriteTest.testInt(intT, max);
+	}
+}
+
+test "read/write signed int" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	try ReadWriteTest.testInt(i2, 0);
+	try ReadWriteTest.testInt(i2, 1);
+	try ReadWriteTest.testInt(i2, -1);
+
+	inline for([_]type{i4, i5, i8, i16, i31, i32, i64, i128}) |intT| {
+		const min = std.math.minInt(intT);
+		std.debug.assert(min != 0);
+
+		const lowerMid = std.math.minInt(intT)/2;
+		std.debug.assert(lowerMid != 0);
+
+		const upperMid = std.math.maxInt(intT)/2;
+		std.debug.assert(upperMid != 0);
+
+		const max = std.math.maxInt(intT);
+		std.debug.assert(max != 0);
```

Comment:
What's with the asserts here? Without this pointless assert you could test `i2` (and i1) here in the loop too.